### PR TITLE
Add builder for OAuth1aToken

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
Related https://github.com/line/armeria/pull/2680#discussion_r416371720
The current implementation for building OAuth 1a token isn't convenient.

Modifcations:
- Add `OAuth1aTokenBuilder`.

Result:
- You can now easily create `OAuth1aToken`.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+     */
+    public OAuth1aTokenBuilder version(String version) {
+        this.version = requireNonNull(version, "version");
+        return this;
+    }
+
+    /**
+     * Sets additional (or user-defined) parameters.
+     */
+    public OAuth1aTokenBuilder additionals(Map<String, String> additionals) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java:130]: What happens if:

- a user puts existing properties such as `nonce` using this method?
- a user calls this method twice? Shouldn't we merge them?

When a user calls this method, we're gonna have to check if there's an entry with a known key, and if so, we should delegate to the specific setter method.

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java:130]: That's a good suggestion. 👍

File: core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aToken.java [hunk 2]
```diff
@@ -233,38 +237,39 @@ public String toHeaderValue() {
...
+        appendValue(builder, OAUTH_VERSION, version);
+        for (Entry<String, String> entry : additionals.entrySet()) {
+            builder.append(',');
+            appendValue(builder, entry.getKey(), entry.getValue());
         }
+
         return headerValue = builder.toString();
     }
 
+    private static void appendValue(StringBuilder builder, String key, String value) {
```
[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aToken.java:266]: That's a good suggestion, also. 👍

File: core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aToken.java [hunk 3]
```diff
@@ -86,46 +84,8 @@
...
-                    case OAUTH_VERSION:
-                        builder.version(value);
-                        break;
-                    default:
-                        additionals.put(key, value);
-                }
-            }
-        }
-        return builder.additionals(additionals.build()).build();
+        setParameters(builder, params);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aToken.java:87]: Can this be simplified into `builder().additionals(rnn(params)).build()`?

File: core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java [hunk 4]
```diff
@@ -125,13 +137,60 @@ public OAuth1aTokenBuilder version(String version) {
...
      */
     public OAuth1aTokenBuilder additionals(Map<String, String> additionals) {
-        this.additionals = ImmutableMap.copyOf(requireNonNull(additionals, "additionals"));
+        setParameters(this, requireNonNull(additionals, "additionals"));
         return this;
     }
 
+    private void additionals0(ImmutableMap<String, String> additionals) {
+        this.additionals = additionals;
+    }
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java:150]: - Could this method be inlined?
- Could we merge instead of overwriting the previous additionals?

File: core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java [hunk 5]
```diff
@@ -125,13 +137,60 @@ public OAuth1aTokenBuilder version(String version) {
     }
 
     /**
-     * Sets additional (or user-defined) parameters.
+     * Sets additional parameters. If the specified {@code additionals} has pre-defined parameters,
+     * the parameters are set automatically to this builder and removed from the {@code additionals}.
      */
     public OAuth1aTokenBuilder additionals(Map<String, String> additionals) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java:143]: This method is not only for additional parameters but any parameters including pre-defined ones. Perhaps we should rename it to `putAll()`? We could also add methods like `put(CharSequence, CharSequence)`.

File: docs-client/package-lock.json [hunk 6]
```diff
@@ -3369,6 +3369,16 @@
       "dev": true,
       "optional": true
     },
+    "bindings": {
```
Comment (trustin) [inline docs-client/package-lock.json:3372]: Is this change relevant to this pull request? :thinking:

Comment (minwoox) [reply docs-client/package-lock.json:3372]: Gradle build just keeps changing this file.
I made a separate PR to solve this. https://github.com/line/armeria/pull/2777

File: core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java (hunk 5 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java]: Thanks, added `put()` and `putAll()`. I didn't use `CharSequence` though because I feel it's too much.
If it's necessary, jut let me know. 😉

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java]: SGTM. Thanks, @minwoox

File: core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    @Nullable
+    private String signature;
+    @Nullable
+    private String timestamp;
+    @Nullable
+    private String nonce;
+    @Nullable
+    private String realm;
+    private String version = DEFAULT_OAUTH_VERSION;
+    private final Builder<String, String> additionalsBuilder = ImmutableMap.builder();
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java:61]: nit:
```suggestion
    private final ImmutableMap.Builder<String, String> additionalsBuilder = ImmutableMap.builder();
```

File: core/src/test/java/com/linecorp/armeria/common/auth/OAuth1aTokenTest.java [hunk 8]
```diff
@@ -15,65 +15,51 @@
...
+        final OAuth1aToken token = OAuth1aToken.builder()
+                                               .consumerKey("a")
+                                               .token("b")
+                                               .signatureMethod("c")
+                                               .signature("d")
+                                               .timestamp("0")
+                                               .nonce("f")
+                                               .put("x-others", "g")
+                                               .build();
+        final Builder<String, String> paramsBuilder = ImmutableMap.builder();
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/common/auth/OAuth1aTokenTest.java:37]: Ditto

File: core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+import javax.annotation.Nullable;
+
+import com.google.common.base.Ascii;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMap.Builder;
+
+/**
+ * Builds a new {@link OAuth1aToken}.
+ */
+public final class OAuth1aTokenBuilder {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java:42]: How about adding `toString()`?

File: core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java (hunk 7 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/auth/OAuth1aTokenBuilder.java]: Fixed. 😉

**Comment by trustin** [comment]
Thanks a lot, @minwoox

**Comment by minwoox** [comment]
Thanks for reviewing. 😉
